File: IAIso-v5.0/core/iaiso-ruby/test/core_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "iaiso/audit"
require "iaiso/core"

class CoreTest < Minitest::Test
  def test_enum_wire_values
    assert_equal "init",      IAIso::Core::Lifecycle::INIT
    assert_equal "running",   IAIso::Core::Lifecycle::RUNNING
    assert_equal "escalated", IAIso::Core::Lifecycle::ESCALATED
    assert_equal "released",  IAIso::Core::Lifecycle::RELEASED
    assert_equal "locked",    IAIso::Core::Lifecycle::LOCKED
    assert_equal "ok",        IAIso::Core::StepOutcome::OK
  end

  def test_config_rejects_bad_thresholds
    cfg = IAIso::Core::PressureConfig.new(escalation_threshold: 0.9, release_threshold: 0.5)
    assert_raises(IAIso::Core::ConfigError) { cfg.validate! }
  end

  def test_config_rejects_negative_coefficient
    cfg = IAIso::Core::PressureConfig.new(token_coefficient: -1.0)
    assert_raises(IAIso::Core::ConfigError) { cfg.validate! }
  end

  def test_engine_escalates_on_high_pressure
    sink = IAIso::Audit::MemorySink.new
    clk = IAIso::Core::ScriptedClock.new([0.0, 1.0])
    cfg = IAIso::Core::PressureConfig.new(
      escalation_threshold: 0.5, release_threshold: 0.95,
      dissipation_per_step: 0.0, depth_coefficient: 0.6,
    )
    eng = IAIso::Core::PressureEngine.new(
      config: cfg,
      options: IAIso::Core::EngineOptions.new(execution_id: "e", audit_sink: sink, clock: clk, timestamp_clock: clk),
    )
    outcome = eng.step(IAIso::Core::StepInput.build(depth: 1))
    assert_equal IAIso::Core::StepOutcome::ESCALATED, outcome
    assert_in_delta 0.6, eng.pressure, 1e-9
  end

  def test_engine_locks_after_release
    clk = IAIso::Core::ScriptedClock.new([0.0, 1.0, 2.0])
    cfg = IAIso::Core::PressureConfig.new(
      escalation_threshold: 0.5, release_threshold: 0.9,
      dissipation_per_step: 0.0, depth_coefficient: 1.0, post_release_lock: true,
    )
    eng = IAIso::Core::PressureEngine.new(
      config: cfg,
      options: IAIso::Core::EngineOptions.new(execution_id: "e"),
    )
    eng.step(IAIso::Core::StepInput.build(depth: 1))
    assert_equal IAIso::Core::Lifecycle::LOCKED, eng.lifecycle
    nxt = eng.step(IAIso::Core::StepInput.build(depth: 1))
    assert_equal IAIso::Core::StepOutcome::LOCKED, nxt
  end

  def test_engine_reset_emits_reset_event
    sink = IAIso::Audit::MemorySink.new
    clk = IAIso::Core::ScriptedClock.new([0.0, 1.0, 2.0])
    eng = IAIso::Core::PressureEngine.new(
      config: IAIso::Core::PressureConfig.defaults,
      options: IAIso::Core::EngineOptions.new(execution_id: "e", audit_sink: sink, clock: clk, timestamp_clock: clk),
    )
    eng.step(IAIso::Core::StepInput.build(tokens: 100))
    eng.reset
    kinds = sink.events.map(&:kind)
    assert_includes kinds, "engine.reset"
    assert_equal 0.0, eng.pressure
    assert_equal IAIso::Core::Lifecycle::INIT, eng.lifecycle
  end

  def test_bounded_execution_run_emits_closed
    sink = IAIso::Audit::MemorySink.new
    IAIso::Core::BoundedExecution.run(audit_sink: sink) do |exec|
      exec.record_tokens(100, tag: "x")
    end
    kinds = sink.events.map(&:kind)
    assert_includes kinds, "execution.closed"
  end

  def test_bounded_execution_auto_id_when_empty
    sink = IAIso::Audit::MemorySink.new
    IAIso::Core::BoundedExecution.run(audit_sink: sink) { |_e| }
    refute_empty sink.events
    assert sink.events.first.execution_id.start_with?("exec-")
  end

  def test_record_tool_call_advances
    exec = IAIso::Core::BoundedExecution.start
    out = exec.record_tool_call("search", tokens: 100)
    assert_equal IAIso::Core::StepOutcome::OK, out
    assert exec.snapshot.pressure > 0.0
    exec.close
  end
end

File: IAIso-v5.0/core/iaiso-php/tests/Unit/Core/CoreTest.php
<?php

declare(strict_types=1);

namespace IAIso\Tests\Unit\Core;

use IAIso\Audit\MemorySink;
use IAIso\Core\BoundedExecution;
use IAIso\Core\BoundedExecutionOptions;
use IAIso\Core\ConfigException;
use IAIso\Core\EngineOptions;
use IAIso\Core\Lifecycle;
use IAIso\Core\PressureConfig;
use IAIso\Core\PressureEngine;
use IAIso\Core\ScriptedClock;
use IAIso\Core\StepInput;
use IAIso\Core\StepOutcome;
use PHPUnit\Framework\TestCase;

final class CoreTest extends TestCase
{
    public function testEnumWireValuesMatchSpec(): void
    {
        self::assertSame('init', Lifecycle::Init->value);
        self::assertSame('running', Lifecycle::Running->value);
        self::assertSame('escalated', Lifecycle::Escalated->value);
        self::assertSame('released', Lifecycle::Released->value);
        self::assertSame('locked', Lifecycle::Locked->value);
        self::assertSame('ok', StepOutcome::Ok->value);
    }

    public function testConfigRejectsBadThresholds(): void
    {
        $cfg = PressureConfig::builder()->escalationThreshold(0.9)->releaseThreshold(0.5)->build();
        $this->expectException(ConfigException::class);
        $cfg->validate();
    }

    public function testConfigRejectsNegativeCoefficient(): void
    {
        $cfg = PressureConfig::builder()->tokenCoefficient(-1.0)->build();
        $this->expectException(ConfigException::class);
        $cfg->validate();
    }

    public function testEngineEscalatesOnHighPressure(): void
    {
        $sink = new MemorySink();
        $clk = new ScriptedClock([0.0, 1.0]);
        $cfg = PressureConfig::builder()
            ->escalationThreshold(0.5)->releaseThreshold(0.95)
            ->dissipationPerStep(0.0)->depthCoefficient(0.6)->build();
        $eng = new PressureEngine($cfg, new EngineOptions('e', $sink, $clk, $clk));
        $outcome = $eng->step(new StepInput(depth: 1));
        self::assertSame(StepOutcome::Escalated, $outcome);
        self::assertEqualsWithDelta(0.6, $eng->getPressure(), 1e-9);
    }

    public function testEngineLocksAfterRelease(): void
    {
        $clk = new ScriptedClock([0.0, 1.0, 2.0]);
        $cfg = PressureConfig::builder()
            ->escalationThreshold(0.5)->releaseThreshold(0.9)
            ->dissipationPerStep(0.0)->depthCoefficient(1.0)
            ->postReleaseLock(true)->build();
        $eng = new PressureEngine($cfg, new EngineOptions('e', null, $clk, $clk));
        $eng->step(new StepInput(depth: 1)); // pressure → 1.0 → release → lock
        self::assertSame(Lifecycle::Locked, $eng->getLifecycle());
        $next = $eng->step(new StepInput(depth: 1));
        self::assertSame(StepOutcome::Locked, $next);
    }

    public function testEngineResetEmitsResetEvent(): void
    {
        $sink = new MemorySink();
        $clk = new ScriptedClock([0.0, 1.0, 2.0]);
        $cfg = PressureConfig::defaults();
        $eng = new PressureEngine($cfg, new EngineOptions('e', $sink, $clk, $clk));
        $eng->step(new StepInput(tokens: 100));
        $eng->reset();
        $kinds = array_map(fn($e) => $e->kind, $sink->events());
        self::assertContains('engine.reset', $kinds);
        self::assertSame(0.0, $eng->getPressure());
        self::assertSame(Lifecycle::Init, $eng->getLifecycle());
    }

    public function testBoundedExecutionRunEmitsClosed(): void
    {
        $sink = new MemorySink();
        BoundedExecution::run(
            new BoundedExecutionOptions(executionId: 'e1', auditSink: $sink),
            function ($ex): void { $ex->recordTokens(100, 'x'); },
        );
        $kinds = array_map(fn($e) => $e->kind, $sink->events());
        self::assertContains('execution.closed', $kinds);
    }

    public function testBoundedExecutionAutoIdWhenEmpty(): void
    {
        $sink = new MemorySink();
        BoundedExecution::run(
            new BoundedExecutionOptions(auditSink: $sink),
            function ($ex): void {},
        );
        self::assertNotEmpty($sink->events());
        self::assertStringStartsWith('exec-', $sink->events()[0]->executionId);
    }

    public function testRecordToolCallAdvancesEngine(): void
    {
        $exec = BoundedExecution::start(new BoundedExecutionOptions());
        $outcome = $exec->recordToolCall('search', 100);
        self::assertSame(StepOutcome::Ok, $outcome);
        self::assertGreaterThan(0.0, $exec->snapshot()->pressure);
        $exec->close();
    }
}
